Add fungible token information to exchange integration page (#722)
near/docs: docs/roles/integrator/exchange-integration.md

sidebar_label: Exchange Integration
---

### Transaction reference links
 - [Basics](/docs/concepts/transaction)
 - [Specifications](https://nomicon.io/RuntimeSpec/Transactions.html)
 - [Constructing Transactions](/docs/tutorials/create-transactions)

## Native NEAR (Ⓝ)

### Balance Changes

Balance changes on accounts can be tracked by using our [changes endpoint](/docs/api/rpc#view-account-changes). You can test this out by sending tokens to an account using [NEAR CLI](/docs/tools/near-cli).

> - First, make sure you have keys to your account locally. The typical workflow is to set up an account at https://wallet.testnet.near.org, then run the NEAR CLI command `login`. (Example: `near login`)

Author: Weird that I can't edit this, but we're missing the closing three-backticks, yes?

> - First, make sure you have keys to your account locally. The typical workflow is to set up an account at https://wallet.testnet.near.org, then run the NEAR CLI command `login`. (Example: `near login`)

Reviewer: hmmm.. its there.. just not showing
<img width="674" alt="Screen Shot 2021-05-27 at 9 28 19 PM" src="https://user-images.githubusercontent.com/58483342/119929548-98566180-bf32-11eb-8645-7f09f4b65c87.png">

> - First, make sure you have keys to your account locally. The typical workflow is to set up an account at https://wallet.testnet.near.org, then run the NEAR CLI command `login`. (Example: `near login`)

Author: Then this GitHub suggestion thing seems borked, good for an actual commit. I'm not gonna merge this, seems a glitch in this suggestion feature thingie

**Note:** Gas prices can change between blocks. Even for transactions with deterministic gas cost, the cost in NEAR could also be different.

### Accounts

Please see the [documentation for accounts](/docs/concepts/account) for basic information.

- For exchanges, NEAR supports [implicit account](https://nomicon.io/DataStructures/Account.html#implicit-account-ids) creation which allows the creation of accounts without paying for transactions.
- You can create an implicit account by following the steps in [this guide](/docs/roles/integrator/implicit-accounts).

> - Accounts must have enough tokens cover its storage. Storage cost per byte is 0.0001 NEAR, and an account with one access key must maintain a balance of at least 0.0182 NEAR. For more details, see [this section of the economics paper](https://near.org/papers/economics-in-sharded-blockchain/#transaction-and-storage-fees).

Author: "typically costs"? I thought this was added by Bowen because it's precise. I think this is a constant, and don't want to add ambiguity

> - Accounts must have enough tokens cover its storage. Storage cost per byte is 0.0001 NEAR, and an account with one access key must maintain a balance of at least 0.0182 NEAR. For more details, see [this section of the economics paper](https://near.org/papers/economics-in-sharded-blockchain/#transaction-and-storage-fees).

Reviewer: Storage price actually can change but will never increase (as it would break a bunch of accounts) but can decrease if the price of NEAR goes up significantly and we want to do another 10x reduction in cost. I should change this wording from "typically" to "currently".

which indicates that the transaction is successful.

## Fungible tokens

Please see the [spec for the fungible token standard](https://nomicon.io/Standards/FungibleToken/README.html) and an [example implementation](https://github.com/near-examples/FT) for reference details.

One notable aspect of the standard is that method names are prefixed with `ft_`. This will be a helpful convention when querying for transactions related to fungible tokens.

### Get balance

> Using the abstraction of the NEAR CLI tool, we can check the balance of a user's account with this command:

Reviewer: ```suggestion
Using the abstraction of the [NEAR CLI](/docs/tools/near-cli) tool, we can check the balance of a user's account with [`near view`](/docs/tools/near-cli#near-view):
```

`near view ft.demo.testnet ft_balance_of '{"account_id": "mike.testnet"}'`

Returns:

```
View call: ft.demo.testnet.ft_balance_of({"account_id": "mike.testnet"})
'1000000'
```

> Using HTTPie:

Reviewer: ```suggestion
Alternatively, you can [call a contract function](/docs/api/rpc#call-a-contract-function) using the `query` RPC endpoint. Below is an example using Httpie:
```

Author: Gotta capitalize it properly to be https://httpie.io/ HTTPie 😇

Reviewer: done :)

```bash
http post https://rpc.testnet.near.org jsonrpc=2.0 id=ftbalance method=query \
params:='{
  "request_type": "call_function",
  "finality": "final",
  "account_id": "ft.demo.testnet",
  "method_name": "ft_balance_of",
  "args_base64": "eyJhY2NvdW50X2lkIjogIm1pa2UudGVzdG5ldCJ9"
}'
```

Returns:

```bash
HTTP/1.1 200 OK
Alt-Svc: clear
Via: 1.1 google
access-control-allow-origin:
content-length: 176
content-type: application/json
date: Thu, 27 May 2021 12:53:38 GMT

{
    "id": "dontcare",
    "jsonrpc": "2.0",
    "result": {
        "block_hash": "3mvNHpZAsXiJ6SuHU1mbLVB4iXCfh5i5d41pnkaSoaJ5",
        "block_height": 49282350,
        "logs": [],
        "result": [
            34,
            49,
            48,
            48,
            48,
            48,
            48,
            48,
            34
        ]
    }
}
```

> As mentioned earlier, the `result` is an array of bytes. There are various ways to convert bytes into a more human-readable form, such as the [dtool CLI](https://github.com/guoxbin/dtool#installation).

Reviewer: ```suggestion
As mentioned earlier, the `result` is an array of bytes. There are various ways to convert bytes into a more human-readable form such as the [dtool CLI](https://github.com/guoxbin/dtool#installation).
```

> 

Reviewer: ```suggestion

Example:

```

`dtool a2h '[34,49,48,48,48,48,48,48,34]' | dtool h2s`

Returns:

`"1000000"`

> Note that the fungible token balance of the account `mike.testnet` is `1000000` wrapped in double-quotes. This is because of an issue with JSON serialization. Amounts given in arguments and results must be serialized as Base-10 strings, e.g. "100". This is done to avoid JSON limitation of max integer value of 2**53, which can certainly happen with fungible tokens.

Reviewer: ```suggestion
**Note:** The fungible token balance of the account `mike.testnet` is `1000000` wrapped in double-quotes. This is because of an issue with JSON serialization. Amounts given in arguments and results must be serialized as Base-10 strings, e.g. "100". This is done to avoid JSON limitation of max integer value of 2**53, which can certainly happen with fungible tokens.
```

> ### Simple transfer

Author: For this section, @volovyk-s please follow the instructions (I think you actually wrote this part) from the FT example. https://github.com/near-examples/FT
Create a fungible token, then send to another user. For instance, this will add storage for `mike.testnet` to the FT contract deployed to `ft.demo.testnet`:

```bash
near call ft.demo.testnet storage_deposit '{"account_id": "mike.testnet", "registration_only": true}' --accountId ft.demo.testnet --amount .025
near call ft.demo.testnet ft_transfer '{"receiver_id": "mike.testnet", "amount": "1000000"}' --accountId ft.demo.testnet --amount 0.000000000000000000000001
```

Now `mike.testnet` has `1000000` tokens.
After that last NEAR CLI command of `ft_transfer`, open up explorer (maybe this will help understand visually) and also get the info of the transaction using NEAR CLI's `tx-status` command. The exchanges will likely be doing lower-level RPC calls that return values like what we get from the `tx-status` command.

Look through the big list of the transaction status and determine how exchanges can tell that this was successful. (I think it's something like Status » SuccessValue or something)

Next, run the same kind of command, except make it error. Maybe send more tokens than are allowed, put in the wrong parameters, or send to an account that doesn't exist or isn't registered. 
Do the same thing where you'll inspect the transaction and look for ways exchanges can tell that this failed. (I think it's something like Status » Failed)

Author: Probably makes sense to have subheaders for Successful/Failed transfers, like how I have that below with "Transfer and call"

> ### Transfer and call

Author: Similar to the previous section, we'll need to deploy a fungible token to testnet or use the same one from previously. This time, however, we'll want to add another contract. So make a subaccount or whatever and deploy a contract to it. You will be writing this contract. It will be similar to this:
https://github.com/near/near-sdk-rs/blob/master/examples/fungible-token/test-contract-defi/src/lib.rs#L46-L79

I highlighted the key method here which is important. You can simply copy this whole example, remove everything else except this method, and have it return "1" (spoiler: `PromiseOrValue::Value(U128::from(1))`)

Then use NEAR CLI to use the "transfer and call" method on the fungible token (deployed earlier) to send 10 fungible tokens where the receiving contract is this new contract.

Again look at the transaction in Explorer and using NEAR CLI to determine that 9 fungible tokens were transferred. 
(With NEAR CLI you intended to send 10 tokens, then the contract said, "no thanks, I only need 9 and so I'm returning you 1 of them" and in the end only 9 are transferred.)

Show how the results of the promises in this chain of events can be looked at and determined that 9 are sent. You may have to do some base64 decoding (or maybe it's base58) on some stuff you see from `near tx-status …`

Reference: https://nomicon.io/Standards/FungibleToken/Core.html?search=53#reference-level-explanation

Get the minimum storage required for FT
  - using `near-cli`:

    ```bash
    near view <contract_account_id> storage_balance_bounds`
    ```

    Result:

    ```bash
    View call: ft.demo.testnet.storage_balance_bounds()
    { min: '1250000000000000000000', max: '1250000000000000000000' }
    ```

  - with `JSON RPC` call
      ```bash
      http post https://rpc.testnet.near.org jsonrpc=2.0 id=storagebalancebounds method=query \
      params:='{
          "request_type": "call_function",
          "finality": "final",
          "account_id": "<contract_account_id>",
          "method_name": "storage_balance_bounds",
          "args_base64": ""
      }'
      ```

      Example responce:
      ```bash
      HTTP/1.1 200 OK
      Alt-Svc: clear
      Via: 1.1 google
      access-control-allow-origin: 
      content-length: 357
      content-type: application/json
      date: Wed, 02 Jun 2021 15:42:49 GMT
      
      {
          "id": "storagebalancebounds",
          "jsonrpc": "2.0",
          "result": {
              "block_hash": "Fy3mBqwj5nvUDha3X7G61kmUeituHASEX12oCASrChEE",
              "block_height": 50060878,
              "logs": [],
              "result": [

>                   123,

Author: I think just for readability, let's have all these bytes inline. so like

```
123, 34, 109…
```

instead of this really tall text

                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  34,
                  44,
                  34,
                  109,
                  97,
                  120,
                  34,
                  58,
                  34,
                  49,
                  50,
                  53,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  48,
                  34,
                  125
              ]
          }
      }
      ```

      Decoded result should look simmilar to:

      ```bash
      {
        "min": "1250000000000000000000",
        "max": "1250000000000000000000"
      }
      ```

> If there is not enough deposit for the storage or returned value is `null` - you should deposite more storage with the next command:

Author: *deposit

                }
              }
            },
            {
              "proof": [
                {
                  "hash": "EGC9ZPJHTbmCs3aQDuCkFQboGLBxU2uzrSZMsp8WonDu",
                  "direction": "Right"
                },
                {
                  "hash": "EsBd1n7bDAphA3HY84DrrKd1GP1VugeNiqFCET2S5sNG",
                  "direction": "Right"
                },
                {
                  "hash": "H4q3ByfNB7QH9QEuHN3tcGay7tjhsZwjXx3sq3Vm3Lza",
                  "direction": "Left"
                }
              ],
              "block_hash": "3XMoeEdm1zE64aByFuWCrZaxfbvsjMHRFcL8Wsp95vyt",
              "id": "EAPh8XrBMqm6iuVH5jsfemz4YqUxWsV8Mz241cw5tjvE",
              "outcome": {
                "logs": [

                ],
                "receipt_ids": [

                ],
                "gas_burnt": 0,
                "tokens_burnt": "0",
                "executor_id": "serhii.near",
                "status": {
                  "SuccessValue": ""
                }
              }
            }
          ]
        },
        "id": 125
    }
    ```

You can get the same info later by the transaction hash from the previous call:

>   - using NEAR explorer: https://explorer.near.org

Author: *Explorer

> 
>   - using NEAR explorer: https://explorer.near.org

Reviewer 2: ```suggestion

  - using NEAR Explorer: https://explorer.near.org
```

                "receipt_ids": [

                ],
                "gas_burnt": 0,
                "tokens_burnt": "0",
                "executor_id": "serhii.near",
                "status": {
                  "SuccessValue": ""
                }
              }
            },
            {
              "proof": [
                {
                  "hash": "BR3R7tjziEgXMiHaJ7VuuXCo2yBHB2ZzsoxobPhPjFeJ",
                  "direction": "Left"
                },
                {
                  "hash": "4vHM3fbdNwXGMp9uYzVKB13abEM6qdPUuZ9rfrdsaDzc",
                  "direction": "Left"
                }
              ],
              "block_hash": "F9xNWGhJuYW336f3qVaDDAipsyfpudJHYbmt5in3MeMT",
              "id": "DwLMVTdqv9Z4g9QC4AthTXHqqeJVAH4s1tFXHQYMArW7",
              "outcome": {
                "logs": [

                ],
                "receipt_ids": [

                ],
                "gas_burnt": 0,
                "tokens_burnt": "0",
                "executor_id": "serhii.near",
                "status": {
                  "SuccessValue": ""
                }
              }
            }
          ]
        },
        "id": 125
    }
    ```

> How to check if transfer was sucessfull?

Author: ```suggestion
Was the fungible token transfer successful?
```

    ```

    Result example:

    ```bash
    Scheduling a call: berryclub.ek.near.ft_transfer({"receiver_id": "volovyk.near", "amount": "1"}) with attached 0.000000000000000000000001 NEAR
    Receipt: GDeE3Kv1JHgs71A22NEUbgq55r2Hvcnis8gCMyJtQ2mx
    	Log [berryclub.ek.near]: Transfer 1 from serhii.near to volovyk.near
    Transaction Id 3MkWKbXVP8wyy4pBofELqiE1pwx7ie2v3SKCwaobNcEe
    To see the transaction in the transaction explorer, please open this url in your browser
    https://explorer.mainnet.near.org/transactions/3MkWKbXVP8wyy4pBofELqiE1pwx7ie2v3SKCwaobNcEe
    ''
    ```

  - with JSON RPC call:

      ```bash
      http post https://rpc.testnet.near.org jsonrpc=2.0 id=dontcare method=broadcast_tx_commit \
          params:='["CwAAAHNlcmhpaS5uZWFyAAmQpgZcJM5nMc6f3tqmw/YI4eAvc84ZgsKMRRRzhY/6CQAAAAAAAAARAAAAYmVycnljbHViLmVrLm5lYXLLWPIiUOElkDF3u4hLAMJ0Sjeo1V338pDdHIp70va3ewEAAAACCwAAAGZ0X3RyYW5zZmVyKwAAAHsicmVjZWl2ZXJfaWQiOiJ2b2xvdnlrLm5lYXIiLCJhbW91bnQiOiIxIn0AQHoQ81oAAAEAAAAAAAAAAAAAAAAAAAAA7fDOZQt3zCtdS05Y8XaZFlwO/Gd5wkkNAHShzDiLQXk4Q4ixpraLPMJivs35PZD0gocXl1iGFbQ46NG3VllzCA=="]'

>       ```

Reviewer 3: @mikedotexe I fear this is not helpful as is. I had an idea in my backlog to write the docs about transaction construction, but so far it has not happened. I wonder how you came up with this transaction string... :thinking: 

I am using my `near-cli` (Rust implementation) to construct a transaction. Here is a completely offline version of the command (it does not require any network/internet connection and no other tool but `near-cli` binary itself):

```
$ near-cli construct-transaction \
    offline --block-hash 68GkSsoBuDMEHxELV3k3kqb7z6N8ij11Mj9XX4Wv97EV --nonce 123 \
    sender serhii.near \
    receiver berryclub.ek.near \
    add-action call-function --method-name ft_transfer --args '{"receiver_id": "volovyk.near", "amount": "1"}' --deposit '0 NEAR' --gas '100 TGas' skip \
    sign-private-key \
      --signer-secret-key 'ed25519:5Aj1vumfqLGNQjwsQD93fDzKZmhvBdAyCBkwrC83aKMqdyLHdcVyAFZ4W3Ej7Y4dsFPKNiVM99wfaSZReJSq5qz7' \
      --signer-public-key 'ed25519:8AJzPPSmGpNzGCUqsvz4X8zHomUEMqCM27ihgkKjQysX' \
    display
```

Author: Gotcha, so the way we were dealing with this is to have links to a page in docs where you create transactions.

>To follow this guide, please check the step by step instructions on how to create a transaction first.

Which links to:
https://docs.near.org/docs/tutorials/create-transactions#low-level----create-a-transaction

What do you think?

Reviewer 3: Yeah, providing the link definitely makes sense. Maybe we want to mentions some details about the transaction that needs to be sent in this particular case, which is:

```
Transaction {
        block_hash: `68GkSsoBuDMEHxELV3k3kqb7z6N8ij11Mj9XX4Wv97EV`,
        signer_id: "serhii.near",
        public_key: ed25519:3oN25HE9tf9YUGXQyXCFB2rEtpomijG8ghVV2JvrA4zJ,
        nonce: 123,
        receiver_id: "berryclub.ek.near",
        actions: [
            FunctionCall(
                FunctionCallAction {
                    method_name: ft_transfer,
                    args: `{"receiver_id": "volovyk.near", "amount": "1"}`,
                    gas: 100000000000000,
                    deposit: 0,
                },
            ),
        ],
    }
```

You may format it as JSON and only highlight `signer_id`, `receiver_id`, and `actions`.

Author: Thanks, Frol, I've pushed up changes with this idea in mind. I think it's a great idea.

In the future there will be the possibility to track different or multiple shards.

> - Once the config has been changed you can restart the node it will start syncing new archival data, in the case where you want the full archival history you can just delete the data dir and start the node from scratch syncing full history or use one of the latest backups containing the data directory snapshot which can be copied under the near home dir (default: ~/.near/data).

Reviewer: ```suggestion
- Once the config has been changed you can restart the node it will start syncing new archival data. If you want the full archival history you can simply delete the data dir and start the node from scratch syncing full history. To speed up this process you can use one of the latest backups containing the data directory snapshot which can be copied under the near home dir (default: ~/.near/data).
```
